### PR TITLE
fix(service-worker): allow service worker to handle non root base hrefs

### PR DESCRIPTION
**Pull request by mattlewis92**
This is based on the patch provided by @gkalpak in this comment: https://github.com/angular/angular/issues/25055#issuecomment-408073162 It's currently a WIP as one of the existing tests is failing and I can't work out how to write a test to verify the fix for this PR (I thought I'd got it but it fails even with this patch). @gkalpak could you help me get this over the line? Thanks! 😄 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
The service worker doesn't work when a base href is set. A minimal repro is setup here: https://github.com/mattlewis92/angular-service-worker-issue-repro

Issue Number: https://github.com/angular/angular/issues/30505, https://github.com/angular/angular/issues/25055

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by gkalpak** [comment]
Thx for working on this, @mattlewis92 ❤️

I started looking at the PR, but then had to focus on other things. I will get to it in the next few days (hopefully this week). Just wanted to give you an update. Sorry for the slow turn-around 😞

(In the meantime, feel free to rebase it on master; we have merged some changes that affect `packages/service-worker/` lately iirc :wink:)

**Comment by mattlewis92** [comment]
No worries, thanks @gkalpak! I realised that even with this fix, that it still doesn't work if the domain is different (i.e. we host all our static assets on a cdn which is a different domain from the one where the service worker runs). I have a minimal repro that I can push up to show this, or maybe we can just work on getting a failing test running here.

**Review by gkalpak — COMMENTED** [review]
Finally got around to reviewing this. It's a great start, thx again, @mattlewis92 😍

I think this would be easier to manage (and test), if broken up in smaller pieces:
1. Let's leave the problem you mentioned in https://github.com/angular/angular/pull/31082#issuecomment-507247962 (assets on different domain) for a follow-up PR.
2. Let's split this into separate commits for the separate problems it fixes. AFAICT, there are at least two issues being fixed:
    1. Detecting navigation to `/ngsw/state`, when using `--baseHref`.
    2. Detecting requests for assets-/data-groups, when using `--baseHref=./` (i.e. dynamic base).

So, this could be split up into two commits and then we can look into how to test each one. WDYT, @mattlewis92?

File: packages/service-worker/worker/testing/scope.ts [hunk 1]
```diff
@@ -115,10 +116,12 @@ export class SwTestHarness implements ServiceWorkerGlobalScope, Adapter, Context
   }[] = [];
 
   constructor(
-      private server: MockServerState, readonly caches: MockCacheStorage, private origin: string) {
+      private server: MockServerState, readonly caches: MockCacheStorage,
+      public readonly origin: string) {
```
[the remark this answers is not in the document]
Comment (alfaproject) [reply packages/service-worker/worker/testing/scope.ts]: You can have `readonly` in any access level. `public` is the default access level tho if you omit, but personally I always prefer to be explicit

[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/service-worker/worker/testing/scope.ts]: Yeah, sorry, I meant `readonly` without any access modifier on a constructor parameter implies `public` (which is not the default of constructor parameters). I, personally, prefer always being explicit (that's what I do on my personal projects for example), but the convention on this repo (afaict) is to omit `public` when it is implied.

**Comment by mattlewis92** [comment]
Hey @gkalpak, thanks for reviewing this. Unfortunately it turns out that there was a different issue that we were hitting that was breaking our service worker (the browser was caching scripts, styles and images without CORS headers and then the SW wasn't able to fetch them after as it was getting the cached versions). Unfortunately I can't spend any more company time on fixing the bug this PR fixes as it doesn't actually affect us, but feel free to take over the PR! 😄

**Comment by gkalpak** [comment]
Thx for the update @mattlewis92. I'll put it on my TODO list and will eventually get to it, unless someone beats me to it (🙏 :grin:)
Thx for getting it off the ground 🛩

**Comment by juristr** [comment]
@gkalpak I was just hit by this 😅. Meaning we have an app that might be deployed at the root level as well as in a subfolder, depending on the client where it is being deployed. Just as mentioned in #25055

I was wondering what the state of this PR is. Is there something still missing I might be able to help out to resolve the issue?

**Comment by gkalpak** [comment]
@juristr: Glad to hear you hit th...ehm...I mean...well...ehm... 😅 ...anyway 😇 😛

So, this PR is in the state where it needs the following:
1. ♻️ Rebase on master
2. 💬 Address [review feedback][1].
3. 🔬 Add tests.
4. :shipit: Merge.
5. 💰 Profit.

[1]: https://github.com/angular/angular/pull/31082#pullrequestreview-256034185

**Comment by juristr** [comment]
> Glad to hear you hit th...ehm...I mean...well...ehm... 😅 ...anyway 😇 😛

😛 haha. Well tbh, I had another issue before and now appending `--base-url ./` to my build the SW seems to pick up the paths correctly.

but now I'm actually wondering what this PR introduces 😅, and whether I'm not missing something where my SW could then break again. Right now I have it working with the base-url set. Nevertheless happy to help with this PR (if it is still valuable)

**Comment by gkalpak** [comment]
I am pretty sure that the SW does not work correctly with `--base-href=./` (and there are some issues with subpaths too iirc - for example you can't get to `/ngsw/state`).

I might be wrong, but I would double-check the app if I were you :wink:

**Comment by juristr** [comment]
yea, have deployed it on our staging environment and gonna test it over the next couple of days.  anyway, I’ll give this PR a look and try to help out. Will probably need to wait for after next week though, as I’m going to speak at ngrome and still have a couple of things to prepare 😅. I’ll let u know 👍

**Comment by gkalpak** [comment]
SGTM 👍 Do let us know on how testing on staging went :smiley:
Have fun as ngRome and feel free to ping me when/if you decide to get your hands dirty with this PR. I would be very happy to help/guide you :wink:

**Comment by ShaneCullinane** [comment]
Hi, is this still on the agenda? If I understand the issue here the angular service worker will not work offline when the site is served from www.abc.efg/something - which is what we seem to have run into. We are re-developing an offline app and we'd like to use Angular but this seems to rule it out as an option. Thanks!

**Comment by gkalpak** [comment]
That should work (if configured correctly), @ShaneCullinane. What does not work (and this PR would fix) is `/ngsw/state` and `--base-href=./` (see #25055 and #30505 for more details).

**Comment by ShaneCullinane** [comment]
Thanks for the quick reply, @gkalpak 
I am getting a 504 error (as in #25055) using `--base-href=/abc/`  rather than `--base-href=./`. 
My configuration is certainly the most likely culprit though...

**Comment by sonukapoor** [comment]
I can take a look at this one @gkalpak

**Comment by gkalpak** [comment]
Superceded by #35979. Thx, @mattlewis92, for jump-starting this.

**Pull request closed without merging**